hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/index/IndexResolver.java

```diff
@@ -0,0 +1,1117 @@
+        }
+    }
+
+    /**
+     * Checks whether the index could be used by the engine.
+     * <p>
+     * At the moment only SORTED and HASH indexes could be used.
+     *
+     * @param index the index
+     * @return {@code true} if the index could be used, {@code false} otherwise
+     */
+    private static boolean isIndexSupported(MapTableIndex index) {
+        return index.getType() == IndexType.SORTED || index.getType() == IndexType.HASH;
+    }
+
+    /**
+     * Extracts comparison operands, removing CAST when possible.
+     *
+     * @param node original comparison node
+     * @return a pair of operands
+     */
+    private static BiTuple<RexNode, RexNode> extractComparisonOperands(RexNode node) {
+        assert node instanceof RexCall;
+
+        RexCall node0 = (RexCall) node;
+
+        assert node0.getOperands().size() == 2;
+
+        RexNode operand1 = node0.getOperands().get(0);
+        RexNode operand2 = node0.getOperands().get(1);
+
+        RexNode normalizedOperand1 = removeCastIfPossible(operand1, false);
+        RexNode normalizedOperand2 = removeCastIfPossible(operand2, false);
+
+        return BiTuple.of(normalizedOperand1, normalizedOperand2);
+    }
+
+    /**
+     * Removes CAST operator from the column expression when possible.
+     * <p>
+     * The following expression might be found: {@code CAST(a AS BIGINT) > ?}. This may happen either due to type coercion
+     * during sql->rel conversion, or due to explicit user request. In the general case, a function applied to the column makes
+     * usage of the index on that column impossible.
+     * <p>
+     * In case of the {@code CAST} operator we may try to remove the CAST, thus relying on internal index converters to
+     * downcast the other side of the comparison expression. See {@link TypeConverters}.
+     *
+     * @param node original node, possibly CAST
+     * @param force whether to remove CAST forcefully even for conversion that is otherwise invalid wrt the storage (used for
+     *              {@code IS NULL} operator)
+     * @return original node if there is nothing to unwrap, or the operand of the CAST
+     */
+    private static RexNode removeCastIfPossible(RexNode node, boolean force) {
+        if (node.getKind() == SqlKind.CAST) {
+            RexCall node0 = (RexCall) node;
+
+            RexNode from = node0.getOperands().get(0);
+
+            if (from instanceof RexInputRef) {
+                RelDataType fromType = from.getType();
+                RelDataType toType = node0.getType();
+
+                if (force) {
+                    // Forced unwrap for IS NULL expression
+                    return from;
+                }
+
+                if (fromType.equals(toType)) {
+                    // Redundant conversion => unwrap
+                    return from;
+                }
+
+                if (fromType.getSqlTypeName().getFamily() == NUMERIC && toType.getSqlTypeName().getFamily() == NUMERIC) {
```

Comment:
Doesn't look safe to remove it for any numeric-to-numeric conversion: cast(bigintColumn as tinyint) < 1 should produce an error for values outside of tinyint range.

Reply:
Disallowed invalid CAST unwrapping. Now the type precedence is used to distinguish between allowed and disallowed casts. Also, removed `force` flag from this method because it could lead to the same problem with `IS NULL` condition.
Added tests.

See https://github.com/hazelcast/hazelcast/pull/17382/commits/bf01d4544629c364427a26c94ff0d9e41974378c